src/consensus/encode.rs

```diff
@@ -306,6 +308,24 @@ pub trait Encodable {
     ///
     /// The only errors returned are errors propagated from the writer.
     fn consensus_encode<W: io::Write + ?Sized>(&self, writer: &mut W) -> Result<usize, io::Error>;
+
+    /// Returns the bytes needed when `self` is serialized
+    ///
+    /// The computation could be costly for big objects, see [`Encodable::serialized_len_early_stop`] to limit
+    /// the cost at the expense of precision for big objects
+    fn serialized_len(&self) -> usize {
+        self.consensus_encode(&mut EmptyWrite).unwrap()
+    }
+
+    /// Equivalent to [`Encodable::serialized_len`] wrapped in `Ok` when the needed bytes to serialize are less
+    /// than `threshold`.
+    ///
+    /// If more than `threshold` bytes are needed the result is an `Err`, containing a number that
+    /// could be a minimum bound of the needed bytes to serialize.
+    fn serialized_len_early_stop(&self, threshold: usize) -> Result<usize, usize> {
```

Comment:
> Maybe this could be called reserve_suggestion as I've seen in std somewhere.

`reserve_suggestion` is ok for me, let's see what others think/propose.  

> Also I still fail to understand why counting bytes is better than counting iterations. It seems it shouldn't be.

I think there are advantages here because you can't have an `is_big_object` decent default implementation.

If you default `is_big_object` to `false` like done in #1027 you may forget to consider big objects. For example, `Witness` could be a big object but I forgot to override it. Now that I think more, even worst, `Transaction::is_big_object` is simply wrong because an input could contain a witness with a million elements and still consider it a small object.
If you default `is_big_object` to `true` you haven't gains until you override, and as shown by the `Witness` could be tricky to implement.

Another advantage is that the initial capacity for big objects isn't initialized while here is at least 512, so it saves about 7 re-allocations for objects greater than 512 bytes. For huge objects the percentual performance difference could be negligible but for "barely big" objects could be noticeable.

Reply:
If you base it on bytes you have a similar problem though. With 512 bytes limit a `Vec<TxOut>` with all `script_pubkey` being empty would perform  56 iterations before figuring out it's big.

Not providing a default is a possibility but I think counting the number of `write`/`write_all` calls similarly to how you count bytes now could work well. If `threshold` is `&mut usize` it can be passed around recursively avoiding problem with witness.

Reply:
> If you base it on bytes you have a similar problem though. With 512 bytes limit a Vec<TxOut> with all script_pubkey being empty would perform 56 iterations before figuring out it's big.

Still decent? I think we can even lower the 512 if we want to be conservative from that side

> Not providing a default is a possibility but I think counting the number of write/write_all calls similarly to how you count bytes now could work well. 

That's the worst case even now, agreed, but why not leveraging a better byte count in the good case?

Reply:
You mean counting them along the way and returning? I'd agree with that.

Reply:
I have some more ideas:

The cost of computing some lengths is zero - they are statically known. This is true for all hashes, primitive types, arrays/tuples/structs of types with statically-known length. It'd be best to not count it towards the threshold. So e.g. `Vec<OutPoint>` should have computation cost 1 even if it contains million items.

I think we can take this into account and make things simpler by providing macros. If we have `const IS_ENCODED_LEN_STATICALLY_KNOWN: bool` then `Vec<T>` where `T: IS_ENCODED_LEN_STATICALLY_KNOWN = true` has cost 1 and can use multiplication to compute the lenght. Otherwise it recurses.

For structs we can have a macro that does the correct thing so there's lower chance people will mess it up if they just use the macro.

For things like `Vec<T>`, `Witness` and primitive types we will have to implement it manually, but it seems acceptable to me. We have a pretty good review process. :)

Reply:
I have experimented something similar in this branch https://github.com/rust-bitcoin/rust-bitcoin/compare/master...RCasatta:rust-bitcoin:serialize_hint?expand=1 ...

Reply:
Cool! Any interesting bench results?

Meanwhile I also got the idea that many types consist of statically-known part and statically-unknown part. E.g. `TxIn` and `TxOut` both have variable-length scripts but the rest is known. Instead of iterating over `Vec` and summing all parts we could compute `n * const + sum_script_lens`. Saves a bunch of additions but we would need to measure the difference to determine if it's worth it.

Reply:
I favored this because the most gains are in being precise for small objects which this one achieve better.
Anyway, now I added STATIC_SERIALIZED_LEN to this to leverage advantages on vecs of elements statically known